sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py

```diff
@@ -172,6 +214,60 @@ def on_exception(self, request):
         return
 
 
+class MultitenantCredentialPolicy(BearerTokenCredentialPolicy):
+    """Adds a bearer token Authorization header to requests, for the tenant provided in authentication challenges.
+
+    See https://docs.microsoft.com/azure/active-directory/develop/claims-challenge for documentation on AAD
+    authentication challenges.
+
+    :param credential: The credential.
+    :type credential: ~azure.core.TokenCredential
+    :param str scopes: Lets you specify the type of access needed.
+    :keyword bool discover_tenant: Determines if tenant discovery should be enabled. Defaults to True.
+    :keyword bool discover_scopes: Determines if scopes from authentication challenges should be provided to token
+        requests, instead of the scopes given to the policy's constructor. Defaults to True.
+    :raises: :class:`~azure.core.exceptions.ServiceRequestError`
+    """
+
+    def __init__(
+        self,
+        credential: "TokenCredential",
+        *scopes: str,
+        discover_tenant: bool = True,
+        discover_scopes: bool = True,
+        **kwargs: "Any"
+    ) -> None:
+        self._discover_tenant = discover_tenant
+        self._discover_scopes = discover_scopes
+        super().__init__(credential, *scopes, **kwargs)
+
+    def on_challenge(self, request: "PipelineRequest", response: "PipelineResponse") -> bool:
+        """Authorize request according to an authentication challenge
+
+        This method is called when the resource provider responds 401 with a WWW-Authenticate header.
+
+        :param ~azure.core.pipeline.PipelineRequest request: the request which elicited an authentication challenge
+        :param ~azure.core.pipeline.PipelineResponse response: the resource provider's response
+        :returns: a bool indicating whether the policy should send the request
+        """
+        if not self._discover_tenant and not self._discover_scopes:
+            # We can't discover the tenant or use a different scope; the request will fail because it hasn't changed
+            return False
+
+        try:
+            challenge = _HttpChallenge(response.http_response.headers.get("WWW-Authenticate"))
+            # azure-identity credentials require an AADv2 scope but the challenge may specify an AADv1 resource
+            scope = challenge.scope or challenge.resource + "/.default" if self._discover_scopes else self._scopes
```

Comment:
It looks like there's a possibility the neither `challenge.scope` _or_ `challenge.resource` are populated, in which case I think we'd just be sending "/.default" if `discover_scopes` is true. In this scenario would sending that be the best/correct thing to do and let it fail at the service?
Or should we be falling back to `self._scopes` in the case that the information in the challenge was insufficient?

Reply:
That's a good point -- I was thinking that we could let the invalid scope error during the service request in that case, but I don't think it would be a useful error experience. We would get something like
`azure.core.exceptions.ClientAuthenticationError: DefaultAzureCredential failed to retrieve a token from the included credentials.`
with further details of
`Authentication failed: AADSTS70011: The provided request must include a 'scope' input parameter. The provided value for the input parameter 'scope' is not valid. The scope /.default is not valid.`

That would indicate that there's an issue with scopes, but a user would have to understand what's going on under the hood to figure out that they should set `discover_scopes` to False. I guess the question is, when we detect that `challenge.scope` and `challenge.resource` are empty, should we:
1. Raise an error with a useful message? Or,
2. Default to using `self._scopes`?

Reply:
For now, I'll go with the latter approach (and make this behavior clear in the class docstring)